### PR TITLE
WIP: Adds GPG signing pipeline

### LINKED ISSUE #8: send metrics somewhere?
just to know how many people are actually using it 👍

### LINKED ISSUE #166: Upload signatures for binaries produced
Create and upload signatures for the binaries (or all files) produced.

This will probably involve some kind of config for signing key or method, like PGP for creating the signatures.

### PR DESCRIPTION
**Pull request by simonswine**
This is WiP, (state works for me)

Just like to get ideas that need to be supported in config other than the key id. I can look after tests and and documentation then.

Fixes #166 

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] I have read the **CONTRIBUTING** document.
- [x] `make ci` passes on my machine.
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.

### CODE REVIEW

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: 07ad1ee1ce23499666c013ed5123fe3220824a6b
  * Commits must be in the following format: **%{type}: %{description}**

Guidelines are available at https://github.com/goreleaser/goreleaser/blob/master/CONTRIBUTING.md#create-a-commit

---
This message was auto-generated by https://gitcop.com

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: 07ad1ee1ce23499666c013ed5123fe3220824a6b
  * Commits must be in the following format: **%{type}: %{description}**

Guidelines are available at https://github.com/goreleaser/goreleaser/blob/master/CONTRIBUTING.md#create-a-commit

---
This message was auto-generated by https://gitcop.com

File: goreleaserlib/goreleaser.go [hunk 1]
```diff
@@ -21,25 +21,27 @@ import (
 	"github.com/goreleaser/goreleaser/pipeline/env"
 	"github.com/goreleaser/goreleaser/pipeline/fpm"
 	"github.com/goreleaser/goreleaser/pipeline/git"
+	"github.com/goreleaser/goreleaser/pipeline/gpg_signing"
```
Comment (caarlos0) [inline goreleaserlib/goreleaser.go]: https://blog.golang.org/package-names

having underscores in package names is not a good practice..

File: pipeline/gpg_signing/signing.go [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+	}
+	keyRingFile, err := os.OpenFile(
+		keyRingPath,
+		os.O_RDONLY,
+		0644,
+	)
+	if err != nil {
+		return err
+	}
+	defer keyRingFile.Close()
```
Comment (caarlos0) [inline pipeline/gpg_signing/signing.go]: this could err...

File: pipeline/gpg_signing/signing.go [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+	signatureFilename := fmt.Sprintf("%s.asc", name)
+	signatureFile, err := os.OpenFile(
+		filepath.Join(ctx.Config.Dist, signatureFilename),
+		os.O_APPEND|os.O_WRONLY|os.O_CREATE|os.O_TRUNC,
+		0644,
+	)
+	if err != nil {
+		return "", err
+	}
+	defer signatureFile.Close()
```
Comment (caarlos0) [inline pipeline/gpg_signing/signing.go]: this could err

File: pipeline/gpg_signing/signing.go [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+
+	artifactFile, err := os.OpenFile(
+		filepath.Join(ctx.Config.Dist, name),
+		os.O_RDONLY,
+		0644,
+	)
+	if err != nil {
+		return "", err
+	}
+	defer artifactFile.Close()
```
Comment (caarlos0) [inline pipeline/gpg_signing/signing.go]: also could err

File: pipeline/gpg_signing/signing.go (hunk 2 above)
Comment (magiconair) [reply pipeline/gpg_signing/signing.go]: Yes, but what are you going to do with the error? The operation has already completed successfully. In general, I ignore `file.Close()` errors unless the code depends on the file being closed.

Comment (caarlos0) [reply pipeline/gpg_signing/signing.go]: I usually only `log.WithError(err).Errorf("failed to close file %s", file)`...

**Comment by magiconair** [comment]
I think `gpg_signing` -> `sign` makes sense. 

The current implementation also signs every artifact which isn't necessary. Signing the checksum file is sufficient. However, people may have existing workflows so this should be configurable.

Another thing is that this doesn't seem to work with gpg keys that have a passphrase. 

I'm working on a branch to address this since I need that functionality myself but I also don't want to take the PR away from @simonswine .

**Comment by magiconair** [comment]
FWIW: It took me a bit to realize that the 20 byte fingerprint is not the key id :(

This is how I got the key id:

```
gpg --keyid-format LONG --list-keys
```

I'm using the fingerprint in my `.gitconfig`. so that was a bit confusing. Question is whether it would make more sense to integrate with the gpg program if installed since it would also handle the pin entry.

**Comment by magiconair** [comment]
```yaml
sign:
  gpg_key_id: abcdedeadbeef
  checksum_only: (true|false)
```

I'd make the `checksum_only` behavior the default.

**Comment by magiconair** [comment]
Added #452 as a follow-up.

**Comment by caarlos0** [comment]
closing in favor of #460

**Pull request closed without merging**